resilience4j-hedge/src/main/java/io/github/resilience4j/hedge/internal/HedgeImpl.java

```diff
@@ -0,0 +1,253 @@
+/*
+ *
+ *  Copyright 2021: Matthew Sandoz
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *
+ */
+package io.github.resilience4j.hedge.internal;
+
+import io.github.resilience4j.hedge.Hedge;
+import io.github.resilience4j.hedge.HedgeConfig;
+import io.github.resilience4j.hedge.HedgeMetrics;
+import io.github.resilience4j.hedge.event.*;
+import io.vavr.collection.HashMap;
+import io.vavr.collection.Map;
+import io.vavr.control.Try;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+public class HedgeImpl implements Hedge {
+
+    private static final Logger LOG = LoggerFactory.getLogger(HedgeImpl.class);
+
+    private final String name;
+    private final Map<String, String> tags;
+    private final HedgeConfig hedgeConfig;
+    private final HedgeEventProcessor eventProcessor;
+    private final HedgeMetrics metrics;
+
+    public HedgeImpl(String name, HedgeConfig hedgeConfig) {
+        this(name, hedgeConfig, HashMap.empty());
+    }
+
+    public HedgeImpl(String name, HedgeConfig hedgeConfig,
+                     io.vavr.collection.Map<String, String> tags) {
+        this.name = name;
+        this.tags = Objects.requireNonNull(tags, "Tags must not be null");
+        this.hedgeConfig = hedgeConfig;
+        this.eventProcessor = new HedgeEventProcessor();
+        this.metrics = hedgeConfig.newMetrics();
+        if (metrics.requiresInput()) {
```

Comment:
Why is this code block needed?

Reply:
I was thinking that where a metric does not need input from the actual calls, such as "hedge everything over 500ms" - that it would save time. If you disagree, I can def remove it!